Test failure in test_font_glyph (possible race condition)

Intermittently, I get:

```
====================================================== FAILURES ======================================================
__________________________________________________ test_font_glyph ___________________________________________________

    def test_font_glyph():
        """Test loading glyphs"""
        # try both a vispy and system font
        sys_fonts = set(list_fonts()) - set(_vispy_fonts)
        assert_true(len(sys_fonts) > 0)
        for face in ('OpenSans', list(sys_fonts)[0]):
            font_dict = dict(face=face, size=12, bold=False, italic=False)
            glyphs_dict = dict()
            chars = 'foobar^C&#'
            if face != 'OpenSans' and os.getenv('APPVEYOR', '').lower() == 'true':
                continue  # strange system font failure
            if 'true' in os.getenv('TRAVIS', '') and sys.version[0] == '3':
                continue  # as of April 2015 strange FontConfig error on Travis
            for char in chars:
                # Warning that Arial might not exist
>               _load_glyph(font_dict, char, glyphs_dict)

vispy/util/fonts/tests/test_font.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
vispy/util/fonts/_freetype.py:48: in _load_glyph
    face.set_char_size(f['size'] * 64)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <vispy.ext.freetype.Face object at 0x2af9dc656208>, width = 768, height = 0, hres = 72, vres = 72

    def set_char_size(self, width=0, height=0, hres=72, vres=72):
        error = FT_Set_Char_Size(self._FT_Face, width, height, hres, vres)
        if error:
>           raise RuntimeError('Could not set size: %s' % hex(error))
E           RuntimeError: Could not set size: 0x17

vispy/ext/freetype.py:397: RuntimeError
```

Automating this test 10,000 times passes each time but running it manually (or via the test suite) seems to fail occasionally with the same error - I've repeated it twice.

Ubuntu 14.04 LTS, AwesomeWM (tiling window manager), NVidia GTX 980, Anaconda Python 3.4.

--- a/vispy/ext/fontconfig.py
+++ b/vispy/ext/fontconfig.py
@@ -113,6 +113,6 @@ def find_font(face, bold, italic):
 
 def _list_fonts():
     """List system fonts"""
-    stdout_, stderr = run_subprocess(['fc-list', ':', 'family'])
+    stdout_, stderr = run_subprocess(['fc-list', ':scalable=true', 'family'])
     vals = [v.split(',')[0] for v in stdout_.strip().splitlines(False)]
     return vals

--- a/vispy/ext/freetype.py
+++ b/vispy/ext/freetype.py
@@ -30,6 +30,7 @@ FT_LOAD_NO_AUTOHINT = 32768
 FT_LOAD_TARGET_LCD = 196608
 FT_LOAD_TARGET_LIGHT = 65536
 FT_LOAD_NO_SCALE = 1
+FT_FACE_FLAG_SCALABLE = 1
 
 _64_bit = (8 * struct.calcsize("P")) == 64
 ##############################################################################

--- a/vispy/util/fonts/_freetype.py
+++ b/vispy/util/fonts/_freetype.py
@@ -26,7 +26,7 @@ _font_dict = {}
 # and isn't using fonts (Windows)
 
 def _load_font(face, bold, italic):
-    from ...ext.freetype import Face
+    from ...ext.freetype import Face, FT_FACE_FLAG_SCALABLE
     key = '%s-%s-%s' % (face, bold, italic)
     if key in _font_dict:
         return _font_dict[key]
@@ -35,6 +35,9 @@ def _load_font(face, bold, italic):
     else:
         fname = find_font(face, bold, italic)
     font = Face(fname)
+    if (FT_FACE_FLAG_SCALABLE & font.face_flags) == 0:
+        raise RuntimeError('Font %s is not scalable, so cannot be loaded'
+                           % face)
     _font_dict[key] = font
     return font
 

--- a/vispy/util/fonts/_win32.py
+++ b/vispy/util/fonts/_win32.py
@@ -23,7 +23,7 @@ from ...ext.gdi32plus import (gdiplus, gdi32, user32, winreg, LOGFONT,
 
 # XXX This isn't perfect, but it should work for now...
 
-def find_font(face, bold, italic):
+def find_font(face, bold, italic, orig_face=None):
     style_dict = {'Regular': 0, 'Bold': 1, 'Italic': 2, 'Bold Italic': 3}
 
     # Figure out which font to actually use by trying to instantiate by name
@@ -46,7 +46,7 @@ def find_font(face, bold, italic):
     use_face = cast(byref(metrics, metrics.otmpFamilyName), c_wchar_p).value
     if use_face != face:
         warnings.warn('Could not find face match "%s", falling back to "%s"'
-                      % (face, use_face))
+                      % (orig_face or face, use_face))
     use_style = cast(byref(metrics, metrics.otmpStyleName), c_wchar_p).value
     use_style = style_dict.get(use_style, 'Regular')
     # AK: I get "Standaard" for use_style, which is Dutch for standard/regular
@@ -78,8 +78,8 @@ def find_font(face, bold, italic):
                     assert gdiplus.GdipGetFamilyName(family, buf, 0) == 0
                     assert buf.value == use_face
                     fname = ff
-    if fname is None:
-        raise RuntimeError('Could not find system font %s' % use_face)
+                    break
+    fname = fname or find_font('', bold, italic, face)  # fall back to default
     return fname
 
 

--- a/vispy/util/fonts/tests/test_font.py
+++ b/vispy/util/fonts/tests/test_font.py
@@ -27,17 +27,15 @@ def test_font_glyph():
     # try both a vispy and system font
     sys_fonts = set(list_fonts()) - set(_vispy_fonts)
     assert_true(len(sys_fonts) > 0)
-    for face in ('OpenSans', list(sys_fonts)[0]):
+    for face in ['OpenSans'] + list(sys_fonts):
         font_dict = dict(face=face, size=12, bold=False, italic=False)
         glyphs_dict = dict()
         chars = 'foobar^C&#'
-        if face != 'OpenSans' and os.getenv('APPVEYOR', '').lower() == 'true':
-            continue  # strange system font failure
-        if 'true' in os.getenv('TRAVIS', '') and sys.version[0] == '3':
-            continue  # as of April 2015 strange FontConfig error on Travis
         for char in chars:
             # Warning that Arial might not exist
-            _load_glyph(font_dict, char, glyphs_dict)
+            with warnings.catch_warnings(record=True):
+                warnings.simplefilter('always')
+                _load_glyph(font_dict, char, glyphs_dict)
         assert_equal(len(glyphs_dict), np.unique([c for c in chars]).size)
 
 

--- a/vispy/util/fonts/tests/test_font.py
+++ b/vispy/util/fonts/tests/test_font.py
@@ -2,8 +2,6 @@
 # Copyright (c) 2015, Vispy Development Team.
 # Distributed under the (new) BSD License. See LICENSE.txt for more info.
 
-import os
-import sys
 import numpy as np
 from nose.tools import assert_true, assert_equal
 import warnings
@@ -28,6 +26,7 @@ def test_font_glyph():
     sys_fonts = set(list_fonts()) - set(_vispy_fonts)
     assert_true(len(sys_fonts) > 0)
     for face in ['OpenSans'] + sorted(list(sys_fonts)):
+        print(face)  # useful for debugging
         font_dict = dict(face=face, size=12, bold=False, italic=False)
         glyphs_dict = dict()
         chars = 'foobar^C&#'

--- a/vispy/util/fonts/tests/test_font.py
+++ b/vispy/util/fonts/tests/test_font.py
@@ -27,7 +27,7 @@ def test_font_glyph():
     # try both a vispy and system font
     sys_fonts = set(list_fonts()) - set(_vispy_fonts)
     assert_true(len(sys_fonts) > 0)
-    for face in ['OpenSans'] + list(sys_fonts):
+    for face in ['OpenSans'] + sorted(list(sys_fonts)):
         font_dict = dict(face=face, size=12, bold=False, italic=False)
         glyphs_dict = dict()
         chars = 'foobar^C&#'

